Gc fixes afl part4

bjourne opened the pull request:
Here is another pr to try and fix #1289. I've created a new vocab `compiler.cfg.stacks.padding` which more precisely traces which stack locations are live so you get more accurate gc maps. It supersedes `compiler.cfg.stacks.map`, but I've kept the old one in the history so it's easy to go back to that one if something doesn't work. But it should work. :) You probably need new boot images too. 

I've also written a fix for the "Generic word lookup-base-pointer*..." test failure that the build machines got. And some general improvements to the cpp code too.

mrjbq7 commented on basis/compiler/cfg/builder/blocks/blocks.factor:
```diff
@@ -35,11 +35,12 @@ IN: compiler.cfg.builder.blocks
     [ out-d>> length ] [ in-d>> length ] bi - ;
 
+: emit-call-block ( word height -- )
+    dup adjust-d ##call, make-kill-block ;
```
Isn't this wrong, since its ``( word height -- )``, this calls ``adjust-d`` and ``##call,`` on height?  When it should instead be:

```[ ##call, ] [ adjust-d ] bi* make-kill-block```

?

mrjbq7 commented on basis/compiler/cfg/ssa/destruction/destruction.factor:
```diff
@@ -132,6 +131,7 @@ M: insn cleanup-insn , ;
 PRIVATE>
 
 : destruct-ssa ( cfg -- )
+    H{ } clone leader-map set
```
Does it need to be a ``H{ }`` or can you just call ``leader-map off``?

bjourne replied:
missed that!

mrjbq7 commented on basis/compiler/cfg/stacks/clearing/clearing.factor:
```diff
@@ -1,11 +1,10 @@
 : state>replaces ( state -- replaces )
-    [ stack>vacant ] map first2
-    [ [ <ds-loc> ] map ] [ [ <rs-loc> ] map ] bi* append
+    [ stack>vacant ] map { ds-loc rs-loc } [ swap create-locs ] 2map concat
```
Does the order of this matter, previously it would be all ds's before all rs's.  Now it is ds/rs alternating?

bjourne replied:
It's still ds before rs

mrjbq7 commented on basis/compiler/cfg/stacks/clearing/clearing.factor:
```diff
@@ -20,6 +19,6 @@ IN: compiler.cfg.stacks.clearing
 
 : clear-uninitialized ( cfg -- )
-    [ trace-stack-state ] keep [
+    [ trace-stack-state2 ] keep [
```
Is there a better name for this, or is intended to replace the old one and eventually be just called ``trace-stack-state2``?

mrjbq7 commented on basis/cpu/architecture/architecture.factor:
```diff
@@ -234,7 +234,7 @@ HOOK: %replace-imm cpu ( src loc -- )
 
 HOOK: stack-frame-size cpu ( stack-frame -- n )
-HOOK: %call cpu ( word -- )
+HOOK: %call cpu ( word height -- )
```
Does any implementation of ``#call`` use the height argument? It looks like they always pass ``0`` and always ``drop`` the height?

mrjbq7 commented on vm/collector.hpp:
```diff
@@ -97,18 +100,12 @@ template <typename TargetGeneration, typename Policy> struct collector {
-
-      start = std::max(start, card_start);
-      end = std::min(end, card_end);
+    start += sizeof(cell);
```
Is ``card_start < end`` always true?

bjourne replied:
No, but if `card_start >= end` then it must be that `start >= end` and then `visit_object_array` won't visit anything.

bjourne commented:
Fixed some small things. The `emit-call-block` is right because I've added a height slot to `##call`. You need it to know how much a call increases or decreases the datastack. It's not needed in `%call` as you noted, but the parameters given to the %-word needs to be the same as the slots in the ##-tuple due to some metaprogramming. Yes the idea is that `trace-stack-state2` will become `trace-stack-state` if the new scheme works as it should.

mrjbq7 commented:
Okay, merged! Thanks!!

bjourne commented:
Damnit. Something seem to break when you run the benchmarks. I'll make a small patch to go back to the old stack tracing.

mrjbq7 commented:
What issue are you seeing with the benchmarks?

bjourne commented:
`benchmark.terrain-generation` crashes if you compile factor in debug mode.
